### ISSUE TITLE
Charts 3.0 - Error: this application, or a library it uses, has passed an invalid numeric value (NaN, or not-a-number) to CoreGraphics API and this value is being ignored

### ISSUE THREAD

**Issue by StevenMasini** [danielgindi/Charts#1626]
Hi Charts Team,

I'm currently updating my project from Switch 2.2 to 3.0 using the master branch of Charts via Cocoapods.

So far I don't meet big issues to update, but I just wanted to report these warnings from my console logs. They are emitted from the Charts framework, and even if they are not blocking me in my update, I thought it would be useful for you if I report them.

`<Error>: Error: this application, or a library it uses, has passed an invalid numeric value (NaN, or not-a-number) to CoreGraphics API and this value is being ignored. Please fix this problem.
If you want to see the backtrace, please set CG_NUMERICS_SHOW_BACKTRACE environmental variable.`

`<Error>: If you want to see the backtrace, please set CG_NUMERICS_SHOW_BACKTRACE environmental variable.
 CGContextAddLineToPoint: no current point.
<Error>: CGContextAddLineToPoint: no current point.`

PS: If you are aware of these warning you can just close this issue.

Good luck with portage, your framework is awesome! 👍

**Comment by danielgindi** [comment]
We know about these - they happen sometimes when the matrix is invalid. (i.e the min and max x/y axis are equals so the range is 0, or an invalid zoom has been passed).
It's always nice to clear those out, so if you find a specific situation which causes a warning please raise an issue with information or create a PR :-)

**Comment by StevenMasini** [comment]
@danielgindi I found out where this warning are coming from.

These errors are generated when my LineChartView try to draw the highlight bar after I set its position manually.

In my code I used to position the highlight bar on the highest value in the chart right after setting up the LineChartData in my LineChartView.

But with Charts 3.0 the highlight bar doesn't draw at all, after I set its position with
`self.chartView.highlightValue(x: Double, dataSetIndex: Int, callDelegate: Bool)`

I succeed to reproduce this errors in the demo by adding this line 
`[_chartView highlightValueWithX:0.0 dataSetIndex:0 callDelegate:true];`
at the end of the function 
`- (void)setDataCount:(int)count range:(double)range`
in the LineChart1ViewController example.

Should I open another issue for this ? or am I doing something wrong here ?
I will investigate further to see if I can localize where the issue come from and maybe submit a PR with the correction.

**Comment by danielgindi** [comment]
We currently have a problem with the manual call to `highlightValue`, as the highlighter now actually expects Y values too. Working on a solution for this...

Thanks for the info! :-)

**Comment by StevenMasini** [comment]
No problem, if I can be of any help let me know.

Good luck with the remain work! ;-)

**Comment by StevenMasini** [comment]
@danielgindi I saw you recently officially release the version 3.0 on Cocoapods. 
But this bug is still present in the current version.

I know you are quite submerged with all the issues reported recently. 
I just would like to know if there is any chance to see this issue corrected ?

Thanks you and good luck :-)

**Comment by StevenMasini** [comment]
@danielgindi Ok never mind my bad. 
I didn't understand your advise last time. You actually need to specify the x and y value when you want to position the highlight bar.

Then why don't you just remove the method that only require the x value ?
`func highlightValue(x: Double, dataSetIndex: Int, callDelegate: Bool)`

I tried it again, it never works. Instead I need to use `func highlightValue(x: Double, y: Double, dataSetIndex: Int, callDelegate: Bool)` to see it working.

**Comment by liuxuan30** [comment]
seems not on tracking - @StevenMasini this still happens right? Can I reproduce it with ChartsDemo? Feel free to reopen it if it still exists

**Comment by StevenMasini** [comment]
@liuxuan30 Yes the problem is still here. As I specified previously, you can reproduce it by:

> I succeed to reproduce this errors in the demo by adding this line 
`[_chartView highlightValueWithX:0.0 dataSetIndex:0 callDelegate:true];`
at the end of the function 
`- (void)setDataCount:(int)count range:(double)range` in the LineChart1ViewController example.

I also start to inspect the code on my own, and it seems this bug occur because of the line 499 in ChartViewBase.swift. 
```swift
open func highlightValue(x: Double, dataSetIndex: Int, callDelegate: Bool)
    {
        highlightValue(x: x, y: Double.nan, dataSetIndex: dataSetIndex, callDelegate: callDelegate)
    }
```

The value `Double.nan` make the error occur. In my own project I succeed to bypass this issue by giving the `y` value to `open func highlightValue(x: Double, y: Double, dataSetIndex: Int, callDelegate: Bool)`.

The solution would be to retrieve the `y` value automatically. I could help with this issue if you want to ? I can correct it in my own branch and submit a pull request.

**Comment by liuxuan30** [comment]
you are more than welcome to file a PR for this :)

**Comment by danielgindi** [comment]
Passing `Double.nan` to Highlight is correct, and does not result in a `NaN` being passed to CoreGraphics. It is just for comparison where looking for the actual value to highlight. `nan` means "ignore the y value, look for any matching value by x-value only".

The problem originates somewhere else. You can put `CG_NUMERICS_SHOW_BACKTRACE` in the ENv, as the warning says, to see a backtrace and pin-point the location.

**Comment by StevenMasini** [comment]
Yes sorry my bad. I should have say "my guess is" instead of `make`, but still it's a wrong guess.

But still the method is working properly if I pass the `y` value manually, so could it be due to a wrong `y` value retrieved ?

I will check the backtrace generated by CG_NUMERICS_SHOW_BACKTRACE to see if I can learn more about this issue.

**Comment by dimohamdy** [comment]
i face this bug

**Comment by mariaCP** [comment]
I confirm this bug; any news about the resolution?

**Comment by Rj707** [comment]
@danielgindi Sir have you solved the below issue ??
**We currently have a problem with the manual call to highlightValue, as the highlighter now actually expects Y values too. Working on a solution for this...**
I m unable to still manually call the highlightValue Method, as i m receiving Nan for both X and Y at delegate method. it works on touch, i mean on uigesturerecognizer touch.

**Comment by gedzionsas** [comment]
I still face this bug with highlightValue

**Comment by jjatie** [comment]
There is one fix in #2568 which will be released with 3.1. Any related issues should be opened in a new issue.

**Comment by liuxuan30** [comment]
I don't think this can be solved in #2568, as it just bypass the NaN. What's causing NaN is ignored.
As this issue does not give enough stack trace or numbers why there is NaN

btw we recently merged #2377, not sure if it helps.
